pkg/controller/servicebindingrequest/binding.go

```diff
@@ -0,0 +1,322 @@
+	sbr *v1alpha1.ServiceBindingRequest,
+	sbrStatus *v1alpha1.ServiceBindingRequestStatus,
+) (
+	*v1alpha1.ServiceBindingRequest,
+	error,
+) {
+	// do not update if both statuses are equal
+	if result := cmp.DeepEqual(sbr.Status, sbrStatus)(); result.Success() {
+		return sbr, nil
+	}
+
+	// coping status over informed object
+	sbr.Status = *sbrStatus
+
+	// converting object into unstructured
+	u, err := converter.ToUnstructured(sbr)
+	if err != nil {
+		return nil, err
+	}
+
+	gr := v1alpha1.SchemeGroupVersion.WithResource(ServiceBindingRequestResource)
+	resourceClient := b.DynClient.Resource(gr).Namespace(sbr.GetNamespace())
+	u, err = resourceClient.UpdateStatus(u, v1.UpdateOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, sbr)
+	if err != nil {
+		return nil, err
+	}
+	return sbr, nil
+}
+
+// onError comprise the update of ServiceBindingRequest status to set error flag, and inspect
+// informed error to apply a different behavior for not-founds.
+func (b *ServiceBinder) onError(
+	err error,
+	sbr *v1alpha1.ServiceBindingRequest,
+	sbrStatus *v1alpha1.ServiceBindingRequestStatus,
+	objs []*unstructured.Unstructured,
+) (reconcile.Result, error) {
+	sbrStatus.BindingStatus = bindingFail
+
+	if objs != nil {
+		sbrStatus.BindingStatus = BindingSuccess
+		b.setApplicationObjects(sbrStatus, objs)
+	}
+
+	_, errStatus := b.updateStatusServiceBindingRequest(sbr, sbrStatus)
+	if errStatus != nil {
+		return RequeueError(errStatus)
+	}
+
+	return RequeueOnNotFound(err, requeueAfter)
+}
+
+// Bind configures binding between the Service Binding Request and its related objects.
+func (b *ServiceBinder) Bind() (reconcile.Result, error) {
+	sbrStatus := b.SBR.Status.DeepCopy()
+
+	b.Logger.Info("Saving data on intermediary secret...")
+	secretObj, err := b.Secret.Commit(b.Data, nil)
+	if err != nil {
+		b.Logger.Error(err, "On saving secret data..")
+		return b.onError(err, b.SBR, sbrStatus, nil)
+	}
+
+	// update status information
+	sbrStatus.BindingStatus = bindingInProgress
+	sbrStatus.Secret = secretObj.GetName()
+
+	updatedObjects, err := b.Binder.Bind()
+	if err != nil {
+		b.Logger.Error(err, "On binding application.")
+		return b.onError(err, b.SBR, sbrStatus, updatedObjects)
+	}
+
+	// saving on status the list of objects that have been touched
+	sbrStatus.BindingStatus = BindingSuccess
+	b.setApplicationObjects(sbrStatus, updatedObjects)
+
+	// annotating objects related to binding
+	namespacedName := types.NamespacedName{Namespace: b.SBR.GetNamespace(), Name: b.SBR.GetName()}
+	if err = SetSBRAnnotations(b.DynClient, namespacedName, b.Objects); err != nil {
+		b.Logger.Error(err, "On setting annotations in related objects.")
+		return b.onError(err, b.SBR, sbrStatus, updatedObjects)
+	}
+
+	// updating status of request instance
+	sbr, err := b.updateStatusServiceBindingRequest(b.SBR, sbrStatus)
+	if err != nil {
+		return RequeueOnConflict(err)
```

Comment:
Shouldn't this be RequeError(err)?

Reply:
Not all errors should be retried, for example a 404. I find it appropriate to only retry on conflicts, since this is the only indication the SBR still exists and can be updated.